File: academic-system/src/main/java/org/example/academic/system/service/AssessmentService.java
package org.example.academic.system.service;

import org.example.academic.system.exception.AcademicSystemException;
import org.example.academic.system.model.AcademicClass;
import org.example.academic.system.model.AcademicSystem;
import org.example.academic.system.model.Assessment;
import org.example.academic.system.model.Assignment;
import org.example.academic.system.model.Exam;
import org.example.academic.system.model.PracticalAssignment;
import org.example.academic.system.model.Seminar;

public class AssessmentService {
    private final AcademicSystem academicSystem;

    public AssessmentService(AcademicSystem academicSystem) {
        this.academicSystem = academicSystem;
    }

    public void registerAssessment(String classCode, String assessmentType, double value, double weight) {
        AcademicClass academicClass = academicSystem.findClassByCode(classCode)
                .orElseThrow(() -> new AcademicSystemException("Class not found."));

        Assessment assessment = createAssessment(assessmentType, value, weight);
        academicClass.addAssessment(assessment);
    }

    private Assessment createAssessment(String assessmentType, double value, double weight) {
        if (assessmentType == null) {
            throw new AcademicSystemException("Assessment type cannot be empty.");
        }

        return switch (assessmentType.trim().toLowerCase()) {
            case "exam" -> new Exam(value, weight);
            case "practical", "practical assignment" -> new PracticalAssignment(value, weight);
            case "seminar" -> new Seminar(value, weight);
            case "assignment" -> new Assignment(value, weight);
            default -> throw new AcademicSystemException("Invalid assessment type.");
        };
    }
}

File: academic-system-python/academic_system/service/assessment_service.py
from ..exception.academic_system_exception import AcademicSystemException
from ..model.academic_system import AcademicSystem
from ..model.assignment import Assignment
from ..model.exam import Exam
from ..model.practical_assignment import PracticalAssignment
from ..model.seminar import Seminar


class AssessmentService:
    def __init__(self, academic_system: AcademicSystem):
        self._academic_system = academic_system

    def register_assessment(
        self, class_code: str, assessment_type: str, value: float, weight: float
    ) -> None:
        academic_class = self._academic_system.find_by_code(class_code)
        if academic_class is None:
            raise AcademicSystemException(f"Turma com codigo '{class_code}' nao encontrada.")
        assessment = self._create_assessment(assessment_type, value, weight)
        academic_class.add_assessment(assessment)

    def _create_assessment(self, assessment_type: str, value: float, weight: float):
        if assessment_type is None:
            raise AcademicSystemException("Tipo de avaliacao nao pode ser nulo.")
        t = assessment_type.lower().strip()
        if t == "exam":
            return Exam(value, weight)
        elif t in ("practical", "practical assignment"):
            return PracticalAssignment(value, weight)
        elif t == "seminar":
            return Seminar(value, weight)
        elif t == "assignment":
            return Assignment(value, weight)
        raise AcademicSystemException(f"Tipo de avaliacao invalido: {assessment_type}")
